client/starwhale/api/_impl/dataset/builder.py

```diff
@@ -48,10 +43,212 @@
 _header_size = _header_struct.size
 _header_version = 0
 
-
 _BDType = t.TypeVar("_BDType", bound="BaseBuildExecutor")
 
 
+class BinWriter(Protocol):
+    total_bin_size: int
+
+    def write_row(self, row: TabularDatasetRow) -> None:
+        """
+        Find large bytes or local fs file in row data. Convert them to accessible link
+        """
+        ...
+
+    def flush(self) -> None:
+        ...
+
+    def __enter__(self) -> BinWriter:
+        return self
+
+    def __exit__(
+        self,
+        type: t.Optional[t.Type[BaseException]],
+        value: t.Optional[BaseException],
+        trace: TracebackType,
+    ) -> None:
+        ...
+
+
+class FragmentBinWriter:
+
+    """
+    bin format:
+        header_magic    uint32  I
+        crc             uint32  I
+        _reserved       uint64  Q
+        size            uint32  I
+        padding_size    uint32  I
+        header_version  uint32  I
+        data_magic      uint32  I --> above 32 bytes
+        data bytes...
+        padding bytes...        --> default 4K padding
+    """
+
+    class _BinSection(t.NamedTuple):
+        offset: int
+        size: int
+        raw_data_offset: int
+        raw_data_size: int
+
+    class _SrcPathSpec(t.NamedTuple):
+        src_path: Path
+        remove_src: bool
+
+    def __init__(
+        self,
+        work_dir: Path,
+        data_output_dir: Path,
+        tabular_dataset: TabularDataset,
+        alignment_bytes_size: int = D_ALIGNMENT_SIZE,
+        volume_bytes_size: int = D_FILE_VOLUME_SIZE,
+        bin2link_threshold: int = D_BIN_TO_LINK_THRESHOLD,
```

Comment:
cloud we remove `bin2link_threshold` argument?

Reply:
> cloud we remove `bin2link_threshold` argument?

removed